File: src/arbitrage/scanner.js
// Arbitrage scanner — detects price inconsistencies across logically related markets
// Strategies:
//   monotonicity  — higher threshold must have ≤ probability than lower threshold
//   basket        — sum of all outcomes in a categorical market must equal ~1.0
//   spread        — YES + NO on a binary market should not compress below threshold

import { getActiveMarkets } from '../services/polymarket/api.js';
import { getDb, all, run }  from '../utils/db.js';
import { logger }           from '../utils/logger.js';
import { CONFIG }           from '../../config.js';

const { FEE_PCT, SLIPPAGE_PCT, ARB } = CONFIG;
const ROUND_TRIP_COST = (FEE_PCT + SLIPPAGE_PCT) * 2; // two legs

// ── Market fetching ───────────────────────────────────────────────────────────

async function fetchMarkets() {
  const markets = [];
  const pageSize = 100;
  let offset = 0;

  while (markets.length < ARB.SCAN_LIMIT) {
    const batch = await getActiveMarkets({ limit: pageSize, offset });
    const list  = Array.isArray(batch) ? batch : (batch?.data ?? []);
    if (!list.length) break;
    markets.push(...list);
    if (list.length < pageSize) break;
    offset += pageSize;
  }

  // Only keep markets with price data and sufficient liquidity
  // The Gamma API bulk response uses outcomePrices / clobTokenIds (JSON strings),
  // NOT a `tokens` array — that field only exists on individual market fetches.
  return markets.filter(m =>
    m.active &&
    !m.closed &&
    (m.outcomePrices || m.clobTokenIds) &&
    parseFloat(m.liquidity ?? 0) >= ARB.MIN_LEG_LIQUIDITY
  );
}

// ── Price extraction ──────────────────────────────────────────────────────────
// The Gamma API bulk response encodes prices as a JSON string array in outcomePrices
// and outcome names as a JSON string array in outcomes.
// e.g. outcomePrices = '["0.72","0.28"]', outcomes = '["Yes","No"]'

function extractPrices(market) {
  const rawOutcomes = market.outcomes      ?? '["Yes","No"]';
  const rawPrices   = market.outcomePrices ?? '[]';

  const outcomes = typeof rawOutcomes === 'string' ? JSON.parse(rawOutcomes) : rawOutcomes;
  const prices   = typeof rawPrices   === 'string' ? JSON.parse(rawPrices)   : rawPrices;

  const result = {};
  outcomes.forEach((outcome, i) => {
    const p = prices[i] != null ? parseFloat(prices[i]) : null;
    result[outcome] = (p !== null && !isNaN(p) && p > 0) ? p : null;
  });
  return result;
}

// ── Grouping helpers ──────────────────────────────────────────────────────────

// Extract a numeric threshold from a market question, e.g.:
//   "Will BTC close above $82,000?"  →  82000
//   "Will ETH be above 3,500 USDC?"  →  3500
function extractThreshold(question) {
  if (!question) return null;
  const matches = question.match(/\$?([\d,]+(?:\.\d+)?)\s*[kKmMbB]?\b/g);
  if (!matches) return null;
  for (const raw of matches) {
    const clean = raw.replace(/[$,]/g, '').trim().toLowerCase();
    const mult  = clean.endsWith('k') ? 1e3 : clean.endsWith('m') ? 1e6 : clean.endsWith('b') ? 1e9 : 1;
    const val   = parseFloat(clean) * mult;
    if (val >= 100) return val; // ignore tiny numbers like days-of-month
  }
  return null;
}

// Normalise a question to a "template" by replacing the threshold number with a placeholder.
// Two questions with the same template are part of the same monotonicity group.
function questionTemplate(question) {
  if (!question) return '';
  return question
    .replace(/\$[\d,]+(?:\.\d+)?(?:\s*[kKmMbB])?/g, '$?') // "$82,000" → "$?"
    .replace(/\b[\d,]{4,}(?:\.\d+)?\b/g, '?')              // bare large numbers
    .toLowerCase()
    .replace(/[^a-z0-9?]+/g, ' ')
    .trim();
}

// ── Strategy: monotonicity ─────────────────────────────────────────────────────
// For a list of related binary markets sorted by ascending threshold:
//   P(YES | threshold_low) ≥ P(YES | threshold_high)
//
// A violation (price_high > price_low) creates a guaranteed arbitrage:
//   Buy YES_low  +  Buy NO_high
//   Cost  = price(YES_low) + (1 − price(YES_high))  < 1.0  (always when violation > 0)
//   Return = 1.0 in ALL three scenarios (BTC > high, between, or < low)
//   Profit = 1 − cost − round_trip_fees

function detectMonotonicity(group) {
  if (group.length < 2) return [];

  const sorted       = [...group].sort((a, b) => a.threshold - b.threshold);
  const opportunities = [];

  for (let i = 0; i < sorted.length - 1; i++) {
    for (let j = i + 1; j < sorted.length; j++) {
      const low  = sorted[i]; // smaller threshold → should have HIGHER YES price
      const high = sorted[j]; // larger  threshold → should have LOWER  YES price

      const yesLow  = low.prices['Yes']  ?? low.prices['YES']  ?? null;
      const yesHigh = high.prices['Yes'] ?? high.prices['YES'] ?? null;
      if (yesLow === null || yesHigh === null) continue;
      if (yesLow <= 0 || yesHigh <= 0) continue;

      const violation = yesHigh - yesLow; // positive = mispricing
      if (violation <= 0) continue;

      const cost           = yesLow + (1 - yesHigh);
      const grossProfit    = 1 - cost;
      const expectedProfit = grossProfit - ROUND_TRIP_COST;
      if (expectedProfit <= 0) continue;

      const confidence = Math.min(1,
        (violation / 0.15) * 0.5 +
        (Math.min(low.liquidity, high.liquidity) / 5000) * 0.3 +
        (violation > 0.05 ? 0.2 : 0)
      );

      opportunities.push({
        strategy:        'monotonicity',
        description:     `YES@${low.threshold} (${(yesLow * 100).toFixed(1)}%) + NO@${high.threshold} (${((1 - yesHigh) * 100).toFixed(1)}%) — violación +${(violation * 100).toFixed(1)}%`,
        expected_profit: expectedProfit,
        confidence,
        legs: [
          { market_id: low.conditionId,  outcome: 'Yes', side: 'buy', price: yesLow,       threshold: low.threshold  },
          { market_id: high.conditionId, outcome: 'No',  side: 'buy', price: 1 - yesHigh,  threshold: high.threshold },
        ],
        market_ids: [low.conditionId, high.conditionId],
      });
    }
  }

  return opportunities;
}

// ── Strategy: basket ──────────────────────────────────────────────────────────
// For multi-outcome markets the sum of all outcome prices should ≈ 1.0
// If sum < BASKET_UNDERPRICED_THRESHOLD: buy all outcomes for guaranteed profit

function detectBasket(market) {
  const outcomes = Object.keys(market.prices);
  if (outcomes.length < 3) return null;

  const validPrices = outcomes.map(o => market.prices[o]).filter(p => p !== null && p > 0);
  if (validPrices.length !== outcomes.length) return null;

  const sum = validPrices.reduce((a, b) => a + b, 0);
  if (sum >= ARB.BASKET_UNDERPRICED_THRESHOLD) return null;

  const grossProfit    = 1 - sum;
  const legs           = outcomes.map(o => ({
    market_id: market.conditionId, outcome: o, side: 'buy', price: market.prices[o],
  }));
  const expectedProfit = grossProfit - ROUND_TRIP_COST * legs.length;
  if (expectedProfit <= 0) return null;

  const confidence = Math.min(1, (grossProfit / 0.10) * 0.6 + (market.liquidity / 10000) * 0.4);

  return {
    strategy:        'basket',
    description:     `Multi-resultado suma=${sum.toFixed(3)} — ${(market.question ?? '').slice(0, 60)}`,
    expected_profit: expectedProfit,
    confidence,
    legs,
    market_ids: [market.conditionId],
  };
}

// ── Strategy: spread anomaly ──────────────────────────────────────────────────
// Binary YES + NO should sum to ≈ 1.0; if sum drops below threshold it means
// both sides are cheap — buy both for a guaranteed return of 1.0

function detectSpread(market) {
  const outcomes = Object.keys(market.prices);
  if (outcomes.length !== 2) return null;

  const [oA, oB]  = outcomes;
  const pA        = market.prices[oA];
  const pB        = market.prices[oB];
  if (pA === null || pB === null || pA <= 0 || pB <= 0) return null;

  const sum = pA + pB;
  if (sum >= ARB.SPREAD_ANOMALY_THRESHOLD) return null;

  const grossProfit    = 1 - sum;
  const expectedProfit = grossProfit - ROUND_TRIP_COST * 2;
  if (expectedProfit <= 0) return null;

  const confidence = Math.min(1, (grossProfit / 0.05) * 0.7 + (market.liquidity / 5000) * 0.3);

  return {
    strategy:        'spread',
    description:     `${oA}(${(pA * 100).toFixed(1)}%) + ${oB}(${(pB * 100).toFixed(1)}%) = ${(sum * 100).toFixed(1)}% — suma anómala`,
    expected_profit: expectedProfit,
    confidence,
    legs: [
      { market_id: market.conditionId, outcome: oA, side: 'buy', price: pA },
      { market_id: market.conditionId, outcome: oB, side: 'buy', price: pB },
    ],
    market_ids: [market.conditionId],
  };
}

// ── Persist results ───────────────────────────────────────────────────────────

function saveOpportunity(db, opp) {
  const now      = Date.now();
  const groupKey = opp.market_ids.sort().join('|');

  run(db,
    `INSERT INTO arb_groups (group_key, strategy, market_ids, detected_at)
     VALUES (?, ?, ?, ?)
     ON CONFLICT(group_key, strategy) DO UPDATE SET
       market_ids  = excluded.market_ids,
       detected_at = excluded.detected_at,
       resolved_at = NULL`,
    [groupKey, opp.strategy, JSON.stringify(opp.market_ids), now]
  );

  const group = all(db,
    `SELECT id FROM arb_groups WHERE group_key = ? AND strategy = ?`,
    [groupKey, opp.strategy]
  )[0];

  run(db,
    `INSERT INTO arb_opportunities
       (group_id, strategy, description, expected_profit, confidence, legs, detected_at)
     VALUES (?, ?, ?, ?, ?, ?, ?)`,
    [group.id, opp.strategy, opp.description, opp.expected_profit, opp.confidence,
     JSON.stringify(opp.legs), now]
  );
}

// ── Main ──────────────────────────────────────────────────────────────────────

export async function scan() {
  logger.info('arb:scan:start');
  const db = await getDb();

  // Expire open opportunities older than 2 hours
  run(db,
    `UPDATE arb_opportunities SET status = 'expired'
     WHERE status = 'open' AND detected_at < ?`,
    [Date.now() - 2 * 3600 * 1000]
  );

  const markets = await fetchMarkets();
  logger.info('arb:scan:markets', { count: markets.length });

  if (!markets.length) {
    logger.info('arb:scan:done', { opportunities: 0 });
    return [];
  }

  // Extract prices from the already-included outcomePrices field (no extra API calls)
  const enriched = markets.map(m => ({
    ...m,
    prices:    extractPrices(m),
    liquidity: parseFloat(m.liquidity ?? 0),
    threshold: extractThreshold(m.question),
  }));

  // ── Diagnostics ──────────────────────────────────────────────────────────
  const withThreshold  = enriched.filter(m => m.threshold !== null).length;
  const binaryMarkets  = enriched.filter(m => Object.keys(m.prices).length === 2);
  const pricedBinary   = binaryMarkets.filter(m => {
    const vals = Object.values(m.prices);
    return vals.every(v => v !== null && v > 0);
  });
  const priceSums      = pricedBinary.map(m => Object.values(m.prices).reduce((a, b) => a + b, 0));
  const minSum         = priceSums.length ? Math.min(...priceSums).toFixed(4) : 'n/a';
  const maxSum         = priceSums.length ? Math.max(...priceSums).toFixed(4) : 'n/a';
  const multiOutcome   = enriched.filter(m => Object.keys(m.prices).length >= 3).length;

  logger.info('arb:scan:enriched', {
    count: enriched.length,
    with_threshold: withThreshold,
    binary_priced: pricedBinary.length,
    multi_outcome: multiOutcome,
    binary_sum_min: minSum,
    binary_sum_max: maxSum,
  });

  // ── Group markets by question template for monotonicity ──────────────────
  const templateGroups = new Map();
  for (const m of enriched) {
    if (m.threshold === null) continue;
    const tmpl = questionTemplate(m.question);
    if (!tmpl) continue;
    if (!templateGroups.has(tmpl)) templateGroups.set(tmpl, []);
    templateGroups.get(tmpl).push(m);
  }

  // Log the groups that have 2+ markets (the ones that get checked for violations)
  const groupsChecked = [...templateGroups.values()].filter(g => g.length >= 2);
  logger.info('arb:scan:monotonicity_groups', {
    total_template_groups: templateGroups.size,
    groups_with_2plus: groupsChecked.length,
    // Show up to 5 sample groups with their thresholds and prices
    samples: groupsChecked.slice(0, 5).map(g =>
      g.map(m => ({ t: m.threshold, yes: Object.values(m.prices)[0]?.toFixed(3) }))
    ),
  });

  const opportunities = [];

  for (const [, group] of templateGroups) {
    if (group.length < 2) continue;
    const opps = detectMonotonicity(group);
    for (const opp of opps) {
      logger.info('arb:opportunity', { strategy: opp.strategy, profit: opp.expected_profit.toFixed(4), desc: opp.description });
      saveOpportunity(db, opp);
      opportunities.push(opp);
    }
  }

  for (const m of enriched) {
    const basket = detectBasket(m);
    if (basket) {
      logger.info('arb:opportunity', { strategy: basket.strategy, profit: basket.expected_profit.toFixed(4) });
      saveOpportunity(db, basket);
      opportunities.push(basket);
    }

    const spread = detectSpread(m);
    if (spread) {
      logger.info('arb:opportunity', { strategy: spread.strategy, profit: spread.expected_profit.toFixed(4) });
      saveOpportunity(db, spread);
      opportunities.push(spread);
    }
  }

  logger.info('arb:scan:done', { opportunities: opportunities.length });
  return opportunities;
}

File: backend/strategies/arbitrage/scanner.py
"""
Arbitrage scanner — detecta inconsistencias de precio entre mercados relacionados.

Estrategias:
  monotonicity — umbral mayor debe tener ≤ probabilidad que umbral menor
  basket       — suma de outcomes en mercado categórico debe ≈ 1.0
  spread       — YES + NO en binario no debe comprimirse bajo threshold
"""

import re
import json
import asyncio
import aiosqlite
from config import CONFIG
from db.connection import fetchall, fetchone, execute
from services.polymarket import get_active_markets, get_midpoint_price
from logger import logger

ROUND_TRIP_COST = (CONFIG.fee_pct + CONFIG.slippage_pct) * 2


async def _fetch_markets() -> list[dict]:
    markets  = []
    page_size = 100
    offset   = 0

    while len(markets) < CONFIG.arb_scan_limit:
        batch = await get_active_markets(limit=page_size, offset=offset)
        if not batch:
            break
        markets.extend(batch)
        if len(batch) < page_size:
            break
        offset += page_size

    return [
        m for m in markets
        if m.get("active")
        and not m.get("closed")
        and (m.get("outcomePrices") or m.get("clobTokenIds"))
        and float(m.get("liquidity") or 0) >= CONFIG.arb_min_leg_liquidity
    ]


def _extract_prices(market: dict) -> dict:
    raw_outcomes = market.get("outcomes") or '["Yes","No"]'
    raw_prices   = market.get("outcomePrices") or "[]"

    outcomes = json.loads(raw_outcomes) if isinstance(raw_outcomes, str) else raw_outcomes
    prices   = json.loads(raw_prices)   if isinstance(raw_prices, str)   else raw_prices

    result = {}
    for i, outcome in enumerate(outcomes):
        p = float(prices[i]) if i < len(prices) and prices[i] is not None else None
        result[outcome] = p if (p is not None and p > 0) else None
    return result


def _extract_threshold(question: str | None) -> float | None:
    if not question:
        return None
    matches = re.findall(r'\$?([\d,]+(?:\.\d+)?)\s*[kKmMbB]?\b', question)
    for raw in matches:
        clean = raw.replace("$", "").replace(",", "").strip().lower()
        mult  = 1e3 if clean.endswith("k") else 1e6 if clean.endswith("m") else 1e9 if clean.endswith("b") else 1
        val   = float(clean.rstrip("kmb")) * mult
        if val >= 100:
            return val
    return None


def _question_template(question: str | None) -> str:
    if not question:
        return ""
    q = re.sub(r'\$[\d,]+(?:\.\d+)?(?:\s*[kKmMbB])?', '?', question)
    q = re.sub(r'\b[\d,]+(?:\.\d+)?\s*[kKmMbB]?\b', '?', q)
    q = q.lower()
    q = re.sub(r'[^a-z?]+', ' ', q)
    return re.sub(r'\s+', ' ', q).strip()


def _slug_base(slug: str | None) -> str | None:
    if not slug:
        return None
    s = re.sub(r'-\d{10,}$', '', slug)
    s = re.sub(r'(-\d+)+$', '', s)
    return s.lower()


def _detect_monotonicity(group: list[dict]) -> list[dict]:
    if len(group) < 2:
        return []

    sorted_group  = sorted(group, key=lambda m: m["threshold"])
    opportunities = []

    for i in range(len(sorted_group) - 1):
        for j in range(i + 1, len(sorted_group)):
            low  = sorted_group[i]
            high = sorted_group[j]

            yes_low  = low["prices"].get("Yes")  or low["prices"].get("YES")
            yes_high = high["prices"].get("Yes") or high["prices"].get("YES")
            if yes_low is None or yes_high is None or yes_low <= 0 or yes_high <= 0:
                continue

            violation = yes_high - yes_low
            if violation <= 0:
                continue

            cost            = yes_low + (1 - yes_high)
            gross_profit    = 1 - cost
            expected_profit = gross_profit - ROUND_TRIP_COST
            if expected_profit <= 0:
                continue

            confidence = min(1.0,
                (violation / 0.15) * 0.5 +
                (min(low["liquidity"], high["liquidity"]) / 5000) * 0.3 +
                (0.2 if violation > 0.05 else 0)
            )

            opportunities.append({
                "strategy": "monotonicity",
                "description": f"YES@{low['threshold']} ({yes_low*100:.1f}%) + NO@{high['threshold']} ({(1-yes_high)*100:.1f}%) — violation +{violation*100:.1f}%",
                "expected_profit": expected_profit,
                "confidence": confidence,
                "legs": [
                    {"market_id": low["conditionId"],  "outcome": "Yes", "side": "buy", "price": yes_low,       "threshold": low["threshold"]},
                    {"market_id": high["conditionId"], "outcome": "No",  "side": "buy", "price": 1 - yes_high, "threshold": high["threshold"]},
                ],
                "market_ids": [low["conditionId"], high["conditionId"]],
            })

    return opportunities


def _detect_basket(market: dict) -> dict | None:
    outcomes = list(market["prices"].keys())
    if len(outcomes) < 3:
        return None

    valid_prices = [market["prices"][o] for o in outcomes if market["prices"].get(o)]
    if len(valid_prices) != len(outcomes):
        return None

    total = sum(valid_prices)
    if total >= CONFIG.arb_basket_underpriced_threshold:
        return None

    legs         = [{"market_id": market["conditionId"], "outcome": o, "side": "buy", "price": market["prices"][o]} for o in outcomes]
    gross_profit = 1 - total
    expected     = gross_profit - ROUND_TRIP_COST * len(legs)
    if expected <= 0:
        return None

    confidence = min(1.0, (gross_profit / 0.10) * 0.6 + (market["liquidity"] / 10000) * 0.4)

    return {
        "strategy": "basket",
        "description": f"Multi-outcome sum={total:.3f} — {(market.get('question') or '')[:60]}",
        "expected_profit": expected,
        "confidence": confidence,
        "legs": legs,
        "market_ids": [market["conditionId"]],
    }


def _detect_spread(market: dict) -> dict | None:
    outcomes = list(market["prices"].keys())
    if len(outcomes) != 2:
        return None

    oa, ob = outcomes
    pa, pb = market["prices"].get(oa), market["prices"].get(ob)
    if pa is None or pb is None or pa <= 0 or pb <= 0:
        return None

    total = pa + pb
    if total >= CONFIG.arb_spread_anomaly_threshold:
        return None

    gross    = 1 - total
    expected = gross - ROUND_TRIP_COST * 2
    if expected <= 0:
        return None

    confidence = min(1.0, (gross / 0.05) * 0.7 + (market["liquidity"] / 5000) * 0.3)

    return {
        "strategy": "spread",
        "description": f"{oa}({pa*100:.1f}%) + {ob}({pb*100:.1f}%) = {total*100:.1f}% — anomalous sum",
        "expected_profit": expected,
        "confidence": confidence,
        "legs": [
            {"market_id": market["conditionId"], "outcome": oa, "side": "buy", "price": pa},
            {"market_id": market["conditionId"], "outcome": ob, "side": "buy", "price": pb},
        ],
        "market_ids": [market["conditionId"]],
    }


async def _save_opportunity(db: aiosqlite.Connection, opp: dict) -> None:
    import time
    now       = int(time.time() * 1000)
    group_key = "|".join(sorted(opp["market_ids"]))

    await execute(db, """
        INSERT INTO arb_groups (group_key, strategy, market_ids, detected_at)
        VALUES (?, ?, ?, ?)
        ON CONFLICT(group_key, strategy) DO UPDATE SET
            market_ids  = excluded.market_ids,
            detected_at = excluded.detected_at,
            resolved_at = NULL
    """, (group_key, opp["strategy"], json.dumps(opp["market_ids"]), now))

    group = await fetchone(db,
        "SELECT id FROM arb_groups WHERE group_key = ? AND strategy = ?",
        (group_key, opp["strategy"]),
    )

    await execute(db, """
        INSERT INTO arb_opportunities
            (group_id, strategy, description, expected_profit, confidence, legs, detected_at)
        VALUES (?, ?, ?, ?, ?, ?, ?)
    """, (
        group["id"], opp["strategy"], opp["description"],
        opp["expected_profit"], opp["confidence"],
        json.dumps(opp["legs"]), now,
    ))


async def scan(db: aiosqlite.Connection) -> list[dict]:
    import time
    logger.info("arb:scan:start")

    await execute(db,
        "UPDATE arb_opportunities SET status = 'expired' WHERE status = 'open' AND detected_at < ?",
        (int(time.time() * 1000) - 2 * 3_600_000,),
    )

    markets = await _fetch_markets()
    logger.info("arb:scan:markets", {"count": len(markets)})
    if not markets:
        logger.info("arb:scan:done", {"opportunities": 0})
        return []

    enriched = [
        {
            **m,
            "prices": _extract_prices(m),
            "liquidity": float(m.get("liquidity") or 0),
            "threshold": _extract_threshold(m.get("question")),
        }
        for m in markets
    ]

    # Agrupar para monotonicity
    mono_groups: dict[str, list] = {}

    def add_to_group(key, market):
        if not key:
            return
        mono_groups.setdefault(key, []).append(market)

    for m in enriched:
        if m["threshold"] is None:
            continue
        tmpl = _question_template(m.get("question"))
        if tmpl:
            add_to_group(f"tmpl:{tmpl}", m)
        sb = _slug_base(m.get("slug"))
        if sb:
            add_to_group(f"slug:{sb}", m)

    # Deduplicar dentro de cada grupo
    for key, group in mono_groups.items():
        seen = set()
        mono_groups[key] = [m for m in group if not (m["conditionId"] in seen or seen.add(m["conditionId"]))]

    opportunities = []

    for group in mono_groups.values():
        if len(group) < 2:
            continue
        for opp in _detect_monotonicity(group):
            logger.info("arb:opportunity", {"strategy": opp["strategy"], "profit": f"{opp['expected_profit']:.4f}"})
            await _save_opportunity(db, opp)
            opportunities.append(opp)

    for m in enriched:
        basket = _detect_basket(m)
        if basket:
            logger.info("arb:opportunity", {"strategy": basket["strategy"], "profit": f"{basket['expected_profit']:.4f}"})
            await _save_opportunity(db, basket)
            opportunities.append(basket)

    # Spread: top 50 por liquidez con precios reales de CLOB
    spread_candidates = sorted(
        [m for m in enriched if len(m["prices"]) == 2],
        key=lambda m: m["liquidity"],
        reverse=True,
    )[:50]

    logger.info("arb:scan:spread_check", {"checking": len(spread_candidates)})

    for m in spread_candidates:
        try:
            raw_token_ids = m.get("clobTokenIds") or "[]"
            token_ids     = json.loads(raw_token_ids) if isinstance(raw_token_ids, str) else raw_token_ids
            outcomes      = list(m["prices"].keys())
            if len(token_ids) < 2:
                continue

            p0, p1 = await asyncio.gather(
                get_midpoint_price(token_ids[0]),
                get_midpoint_price(token_ids[1]),
            )

            clob_prices = {outcomes[0]: p0, outcomes[1]: p1}
            m_clob      = {**m, "prices": clob_prices}
            spread      = _detect_spread(m_clob)
            if spread:
                logger.info("arb:opportunity", {"strategy": spread["strategy"], "profit": f"{spread['expected_profit']:.4f}", "sum": f"{p0+p1:.4f}"})
                await _save_opportunity(db, spread)
                opportunities.append(spread)
        except Exception as err:
            logger.warn("arb:spread_clob_error", {"market": m["conditionId"], "error": str(err)})

    logger.info("arb:scan:done", {"opportunities": len(opportunities)})
    return opportunities
